### PR TITLE
add sockets info to operating-systems.md

### PR DESCRIPTION
**Pull request by zafarella**
#7

### CODE REVIEW

File: topics/operating-systems.md [hunk 1]
```diff
@@ -149,3 +149,18 @@
 - A race condition occurs when 2 or more threads are able to access shared data and they try to change it at the same time. Because the thread scheduling algorithm can swap between threads at any point, you don't know the order at which the threads will attempt to access the shared data. Therefore, the result of the change in data is dependent on the thread scheduling algorithm, i.e. both threads are 'racing' to access/change the data.
 - Often problems occur when one thread does a "check-then-act" (e.g. "check" if the value is X, and then "act" to do something that depends on the value being X) and another thread does something to the value in between the "check" and the "act".
 - In order to prevent race conditions occurring, typically you would put a lock around the shared data to ensure that only one thread can access the data at a time.
+
+## Sockets
+- Types of sockets:
+  - Datagram - Uses UPD for delivering packages. No guranty for order and delivery.
```
Comment (orrsella) [inline topics/operating-systems.md]: > Uses UPD

Did you mean `UDP`, or specifically `UPD`? If the latter, maybe explain (or link) to what `UPD` is.

> delivering packages

Is it packages or packets?

File: topics/operating-systems.md [hunk 2]
```diff
@@ -149,3 +149,18 @@
 - A race condition occurs when 2 or more threads are able to access shared data and they try to change it at the same time. Because the thread scheduling algorithm can swap between threads at any point, you don't know the order at which the threads will attempt to access the shared data. Therefore, the result of the change in data is dependent on the thread scheduling algorithm, i.e. both threads are 'racing' to access/change the data.
 - Often problems occur when one thread does a "check-then-act" (e.g. "check" if the value is X, and then "act" to do something that depends on the value being X) and another thread does something to the value in between the "check" and the "act".
 - In order to prevent race conditions occurring, typically you would put a lock around the shared data to ensure that only one thread can access the data at a time.
+
+## Sockets
+- Types of sockets:
+  - Datagram - Uses UPD for delivering packages. No guranty for order and delivery. 
+  - Stream - Order and delivery of packages guranteed. Usually uses TCP protocol.
```
Comment (orrsella) [inline topics/operating-systems.md]: >Usually uses TCP protocol

"uses" or "used for"? Very different meaning...

File: topics/operating-systems.md [hunk 3]
```diff
@@ -149,3 +149,18 @@
 - A race condition occurs when 2 or more threads are able to access shared data and they try to change it at the same time. Because the thread scheduling algorithm can swap between threads at any point, you don't know the order at which the threads will attempt to access the shared data. Therefore, the result of the change in data is dependent on the thread scheduling algorithm, i.e. both threads are 'racing' to access/change the data.
 - Often problems occur when one thread does a "check-then-act" (e.g. "check" if the value is X, and then "act" to do something that depends on the value being X) and another thread does something to the value in between the "check" and the "act".
 - In order to prevent race conditions occurring, typically you would put a lock around the shared data to ensure that only one thread can access the data at a time.
+
+## Sockets
+- Types of sockets:
+  - Datagram - Uses UPD for delivering packages. No guranty for order and delivery. 
+  - Stream - Order and delivery of packages guranteed. Usually uses TCP protocol. 
+  - Raw - no protocol specified on transport layer.
```
Comment (orrsella) [inline topics/operating-systems.md]: nit: Capitalize `no`

File: topics/operating-systems.md [hunk 4]
```diff
@@ -149,3 +149,18 @@
...
+  - Other - Unix Socket for example as one of IPC.
+- System calls
```
Comment (orrsella) [inline topics/operating-systems.md]: nit: add colon at EOL

File: topics/operating-systems.md [hunk 5]
```diff
@@ -149,3 +149,18 @@
...
+  - bind - assigns a name to an unnamed socket
```
Comment (orrsella) [inline topics/operating-systems.md]: nits (for this and all below lines):
1. Put system call name (i.e. "bind") in code block: `bind`.
2. Capitalize descriptions.

File: topics/operating-systems.md [hunk 6]
```diff
@@ -149,3 +149,18 @@
...
+  - connect - to establish a connection with a server
+  - accept - an actual connection from some client process is waited for by having the server execute the accept system call
+  - send, sendto, recv and recvfrom - are similar to the standard read and write system calls
+  - close - to close a socket.
+
```
Comment (orrsella) [inline topics/operating-systems.md]: Please remove empty lines.

File: topics/operating-systems.md [hunk 7]
```diff
@@ -149,3 +149,18 @@
 - A race condition occurs when 2 or more threads are able to access shared data and they try to change it at the same time. Because the thread scheduling algorithm can swap between threads at any point, you don't know the order at which the threads will attempt to access the shared data. Therefore, the result of the change in data is dependent on the thread scheduling algorithm, i.e. both threads are 'racing' to access/change the data.
 - Often problems occur when one thread does a "check-then-act" (e.g. "check" if the value is X, and then "act" to do something that depends on the value being X) and another thread does something to the value in between the "check" and the "act".
 - In order to prevent race conditions occurring, typically you would put a lock around the shared data to ensure that only one thread can access the data at a time.
+
+## Sockets
+- Types of sockets:
+  - Datagram - Uses UPD for delivering packages. No guranty for order and delivery. 
+  - Stream - Order and delivery of packages guranteed. Usually uses TCP protocol. 
+  - Raw - no protocol specified on transport layer.
+  - Other - Unix Socket for example as one of IPC.
```
Comment (orrsella) [inline topics/operating-systems.md]: Reading the man page, I see:

```
The socket has the indicated type, which specifies the semantics of communication.  Currently defined types are:

           SOCK_STREAM
           SOCK_DGRAM
           SOCK_RAW
```

"Other" doesn't appear here. Are you sure that's a socket type? Doesn't "Raw" include "Other"? (again, I'm no expert in sockets...)

Also, according to the man page "Unix" is a "format" of socket, not a type. Again not sure if that's what you meant here.

File: topics/operating-systems.md [hunk 8]
```diff
@@ -149,3 +149,18 @@
 - A race condition occurs when 2 or more threads are able to access shared data and they try to change it at the same time. Because the thread scheduling algorithm can swap between threads at any point, you don't know the order at which the threads will attempt to access the shared data. Therefore, the result of the change in data is dependent on the thread scheduling algorithm, i.e. both threads are 'racing' to access/change the data.
 - Often problems occur when one thread does a "check-then-act" (e.g. "check" if the value is X, and then "act" to do something that depends on the value being X) and another thread does something to the value in between the "check" and the "act".
 - In order to prevent race conditions occurring, typically you would put a lock around the shared data to ensure that only one thread can access the data at a time.
+
+## Sockets
+- Types of sockets:
```
Comment (orrsella) [inline topics/operating-systems.md]: I would consider adding a line above this with a 1-liner explanation on what a socket is. Like:

A socket is one endpoint of a two-way communication link...

**Comment by zafarella** [comment]
@orrsella  addressed comments

**Comment by orrsella** [comment]
Thanks for contributing this and for addressing comments.

**Pull request merged**